### ISSUE TITLE
change(android):  Web-based popup key longpressses

### ISSUE THREAD

**Issue by darcywong00** [keymanapp/keyman#9573]
From the Sept 2023 team planning meetings, it was decided to have Android match iOS (see #2968) in using KeymanWeb for handling longpress keys.

Pros: maintainability of OSK, also avoids duplicating gestures 
Cons: The longpresses would be limited to the top of the OSK / banner

Affects: Keyman Engine for Android consumers (FV, KAB, etc).

This might also end up resolving open issues specific to Android longpresses
e.g. #9499

**Comment by mcdurdin** [comment]
Note: this means banner will _always_ be visible on Android, no option to switch off banner any more.

**Comment by darcywong00** [comment]
Preliminary implementation
![andorid kmw osk](https://github.com/keymanapp/keyman/assets/7358010/df05897e-45e0-4d04-afc3-3362a19926a3)

shows some TODO styling to be consistent with current popup
![image](https://github.com/keymanapp/keyman/assets/7358010/182ee024-196c-46f9-82ca-7a2fe10fad6a)

Does iOS also limit popup to the banner within the Keyman app?

**Comment by darcywong00** [comment]
> Note: this means banner will _always_ be visible on Android, no option to switch off banner any more.

UX question: Should Android follow iOS and display a Keyman image banner when predictions are disabled? It may take me some time to repurpose the iOS banners to the Android form factors.

**Comment by jahorton** [comment]
I think the plan is to go there eventually, but that's not something iOS currently mandates.  iOS has the "Show Banner" toggle - the banner only shows if predictions are available OR if that toggle is on.

In the screenshot below, the toggle is off and there is no lexical model.

![File (3)](https://github.com/keymanapp/keyman/assets/25213402/80f82ac9-0c7d-44b0-94b3-316dfc9d1d15)

**Comment by mcdurdin** [comment]
> > Note: this means banner will _always_ be visible on Android, no option to switch off banner any more.
> 
> UX question: Should Android follow iOS and display a Keyman image banner when predictions are disabled? It may take me some time to repurpose the iOS banners to the Android form factors.

That needs to be done, yes. The UX for longpress when the banner is not visible is not very good.

**Comment by darcywong00** [comment]
ok. so it sounds like I should have Keyman Engine for Android always display the banner (probably a blank black banner) that the apps can then theme with a custom image banner.

The language settings toggle to disable predictions will stop suggestions from appearing, but the banner remains.

**Comment by mcdurdin** [comment]
> ok. so it sounds like I should have Keyman Engine for Android always display the banner (probably a blank black banner) that the apps can then theme with a custom image banner.

Yes, this should be the same in both Android and iOS, ideally. The banner image could be specified by Android (and iOS)'s keyboard.html CSS, and if that's possible, then I suggest that it could be implemented something like this:

```
  #keymanEmptyBanner {  /* I don't know what the appropriate selector is though */
    background-color: #ffffff;
    background-image: url('keyman-banner.png');
    background-size: auto 100%;
    background-repeat: no-repeat;
    background-position: top left;
  }
```

This would scale the image according to the height of the banner, and fill remaining space white, I think. That should handle all the various aspect ratios. One could use keyman-banner.svg instead, which would give a better result on all devicess.

That way we don't need to have a new API to define the image. But this depends on how the existing banner is implemented.

@jahorton thoughts? Does this sound sensible?

**Comment by jahorton** [comment]
What the iOS app actually does is use the `ImageBanner`, passing along a dynamically-constructed data URL for the image's source.

For an interactive version of the idea: https://elmah.io/tools/base64-image-encoder/ can be used to construct a data URL; set the image banner's image-source to the URL spit out and you've got an image.

For a small black box that could be tiled as the background:

```
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABoAAAARCAIAAABM7ytaAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAAAAVSURBVDhPYxgFo2AUjIJRQDpgYAAABT8AAcEGbxwAAAAASUVORK5CYII=" />
```

For what iOS uses, https://github.com/keymanapp/keyman/blob/master/ios/keyman/Keyman/SWKeyboard/ImageBannerViewController.swift has the most compact, centralized version of the code.  It actually renders a full iOS view to an image, then converts _that_ to base64.  See `renderAsBase64` - there were built-in methods on the iOS image type we were able to utilize.

This approach was taken partly due to iOS dark-mode support.

**Comment by jahorton** [comment]
We already use that for the OSK CSS.

**Comment by mcdurdin** [comment]
> Ah right, the other part of the reason we took this approach: it's quite difficult to construct a sufficient set of banner images for all the different aspect ratios supported by iOS; by constructing it at runtime, we can ensure the banner's layout always looks good and appropriate. But yeah, there is an API for this in the iOS engine.

Hence my suggestion of the `background` CSS properties -- that way AFAICT we only need 1 image and aspect ratio issues basically go away.

**Comment by jahorton** [comment]
As a reminder of the app's history, here's how we _used_ to specify the banner's image before predictive text:

https://github.com/keymanapp/keyman/tree/stable-11.0/ios/keyman/Keyman/SWKeyboard/Keyman%20Banner

There were quite a few image files for the various aspect ratios and resolution levels involved.

#1703 seems to be when the change was made.

**Comment by darcywong00** [comment]
https://github.com/keymanapp/keyman/blob/e56c5ba80032101adf618e32495ff28861cbc3f9/web/src/engine/osk/src/banner/bannerView.ts#L176-L190

afaict, even though I have Keyman for Android change the banner to image, KeymanWeb already initialized with a constructor for `new BlankRanner()`.

**Comment by darcywong00** [comment]
Fixed by #9591 and #9696
